File: test/ModularPipelines.UnitTests/Dependencies/DependsOnModulesWithAttributeAttributeTests.cs
using ModularPipelines.Attributes;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Dependencies;

/// <summary>
/// Unit tests for <see cref="DependsOnModulesWithAttributeAttribute{TAttribute}"/>.
/// </summary>
public class DependsOnModulesWithAttributeAttributeTests : TestBase
{
    [Test]
    public async Task ShouldDependOn_ModuleHasAttribute_ReturnsTrue()
    {
        var attr = new DependsOnModulesWithAttributeAttribute<CriticalAttribute>();
        var context = new MockDependencyContext();

        var result = attr.ShouldDependOn(typeof(CriticalModule), context);

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task ShouldDependOn_ModuleLacksAttribute_ReturnsFalse()
    {
        var attr = new DependsOnModulesWithAttributeAttribute<CriticalAttribute>();
        var context = new MockDependencyContext();

        var result = attr.ShouldDependOn(typeof(NonCriticalModule), context);

        await Assert.That(result).IsFalse();
    }

    [Test]
    public async Task ShouldDependOn_ModuleHasInheritedAttribute_ReturnsTrue()
    {
        var attr = new DependsOnModulesWithAttributeAttribute<InheritableAttribute>();
        var context = new MockDependencyContext();

        var result = attr.ShouldDependOn(typeof(DerivedModuleWithInheritedAttribute), context);

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task ShouldDependOn_ModuleHasDifferentAttribute_ReturnsFalse()
    {
        var attr = new DependsOnModulesWithAttributeAttribute<CriticalAttribute>();
        var context = new MockDependencyContext();

        var result = attr.ShouldDependOn(typeof(ModuleWithDifferentAttribute), context);

        await Assert.That(result).IsFalse();
    }

    [Test]
    public async Task ShouldDependOn_ModuleHasMultipleAttributesIncludingMatch_ReturnsTrue()
    {
        var attr = new DependsOnModulesWithAttributeAttribute<CriticalAttribute>();
        var context = new MockDependencyContext();

        var result = attr.ShouldDependOn(typeof(ModuleWithMultipleAttributes), context);

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task ShouldDependOn_CheckingForSerializableAttribute_ReturnsTrue()
    {
        var attr = new DependsOnModulesWithAttributeAttribute<SerializableAttribute>();
        var context = new MockDependencyContext();

        var result = attr.ShouldDependOn(typeof(SerializableModule), context);

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task ShouldDependOn_CheckingForSerializableAttribute_ReturnsFalse()
    {
        var attr = new DependsOnModulesWithAttributeAttribute<SerializableAttribute>();
        var context = new MockDependencyContext();

        var result = attr.ShouldDependOn(typeof(NonCriticalModule), context);

        await Assert.That(result).IsFalse();
    }

    #region Test Attributes and Modules

    [AttributeUsage(AttributeTargets.Class)]
    private sealed class CriticalAttribute : Attribute { }

    [AttributeUsage(AttributeTargets.Class)]
    private sealed class OtherAttribute : Attribute { }

    [AttributeUsage(AttributeTargets.Class, Inherited = true)]
    private sealed class InheritableAttribute : Attribute { }

    [Critical]
    private class CriticalModule { }

    private class NonCriticalModule { }

    [Inheritable]
    private class BaseModuleWithInheritableAttribute { }

    private class DerivedModuleWithInheritedAttribute : BaseModuleWithInheritableAttribute { }

    [Other]
    private class ModuleWithDifferentAttribute { }

    [Critical]
    [Other]
    private class ModuleWithMultipleAttributes { }

    [Serializable]
    private class SerializableModule { }

    #endregion
}

File: test/ModularPipelines.UnitTests.FSharp/Dependencies/DependsOnModulesWithAttributeAttributeTests.fs
namespace ModularPipelines.UnitTests.FSharp.Dependencies

open System
open ModularPipelines.Attributes
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

[<AttributeUsage(AttributeTargets.Class)>]
type CriticalAttribute() =
    inherit Attribute()

[<AttributeUsage(AttributeTargets.Class)>]
type OtherAttribute() =
    inherit Attribute()

[<AttributeUsage(AttributeTargets.Class, Inherited = true)>]
type InheritableAttribute() =
    inherit Attribute()

[<Critical>]
type CriticalModule() =
    class
    end

type NonCriticalModule() =
    class
    end

[<Inheritable>]
type BaseModuleWithInheritableAttribute() =
    class
    end

type DerivedModuleWithInheritedAttribute() =
    inherit BaseModuleWithInheritableAttribute()

[<Other>]
type ModuleWithDifferentAttribute() =
    class
    end

[<Critical>]
[<Other>]
type ModuleWithMultipleAttributes() =
    class
    end

[<Serializable>]
type SerializableModule() =
    class
    end

type DependsOnModulesWithAttributeAttributeTests() =
    [<Test>]
    member _.ShouldDependOn_ModuleHasAttribute_ReturnsTrue() = async {
        let attr = DependsOnModulesWithAttributeAttribute<CriticalAttribute>()
        let context = MockDependencyContext()
        let result = attr.ShouldDependOn(typeof<CriticalModule>, context)
        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.ShouldDependOn_ModuleLacksAttribute_ReturnsFalse() = async {
        let attr = DependsOnModulesWithAttributeAttribute<CriticalAttribute>()
        let context = MockDependencyContext()
        let result = attr.ShouldDependOn(typeof<NonCriticalModule>, context)
        do! check(Assert.That(result).IsFalse())
    }

    [<Test>]
    member _.ShouldDependOn_ModuleHasInheritedAttribute_ReturnsTrue() = async {
        let attr = DependsOnModulesWithAttributeAttribute<InheritableAttribute>()
        let context = MockDependencyContext()
        let result = attr.ShouldDependOn(typeof<DerivedModuleWithInheritedAttribute>, context)
        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.ShouldDependOn_ModuleHasDifferentAttribute_ReturnsFalse() = async {
        let attr = DependsOnModulesWithAttributeAttribute<CriticalAttribute>()
        let context = MockDependencyContext()
        let result = attr.ShouldDependOn(typeof<ModuleWithDifferentAttribute>, context)
        do! check(Assert.That(result).IsFalse())
    }

    [<Test>]
    member _.ShouldDependOn_ModuleHasMultipleAttributesIncludingMatch_ReturnsTrue() = async {
        let attr = DependsOnModulesWithAttributeAttribute<CriticalAttribute>()
        let context = MockDependencyContext()
        let result = attr.ShouldDependOn(typeof<ModuleWithMultipleAttributes>, context)
        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.ShouldDependOn_CheckingForSerializableAttribute_ReturnsTrue() = async {
        let attr = DependsOnModulesWithAttributeAttribute<SerializableAttribute>()
        let context = MockDependencyContext()
        let result = attr.ShouldDependOn(typeof<SerializableModule>, context)
        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.ShouldDependOn_CheckingForSerializableAttribute_ReturnsFalse() = async {
        let attr = DependsOnModulesWithAttributeAttribute<SerializableAttribute>()
        let context = MockDependencyContext()
        let result = attr.ShouldDependOn(typeof<NonCriticalModule>, context)
        do! check(Assert.That(result).IsFalse())
    }
